desktop_server/src/com/mygdx/game/server/ecs/system/VisibilitySystem.java

```diff
@@ -0,0 +1,81 @@
+package com.mygdx.game.server.ecs.system;
+
+import com.artemis.ComponentMapper;
+import com.artemis.EntitySubscription;
+import com.artemis.annotations.All;
+import com.artemis.annotations.AspectDescriptor;
+import com.artemis.systems.IteratingSystem;
+import com.badlogic.gdx.utils.Bits;
+import com.badlogic.gdx.utils.IntMap;
+import com.mygdx.game.core.ecs.component.Coordinates;
+import com.mygdx.game.core.ecs.component.SubField;
+import com.mygdx.game.server.ecs.component.ChangeSubscribers;
+import com.mygdx.game.server.ecs.component.SightlineSubscribers;
+import lombok.extern.java.Log;
+
+import javax.inject.Inject;
+
+@All({Coordinates.class, SightlineSubscribers.class})
+@Log
+public class VisibilitySystem extends IteratingSystem {
+
+  @AspectDescriptor(all = {Coordinates.class, ChangeSubscribers.class}, exclude = { SubField.class })
+  private EntitySubscription allThatCanBePerceived;
+  private final IntMap<Bits> entityToNewChangeSubscribers = new IntMap<>();
+  private ComponentMapper<SightlineSubscribers> sightlineSubscribersMapper;
+  private ComponentMapper<ChangeSubscribers> changeSubscribersMapper;
+  private ComponentMapper<Coordinates> coordinatesMapper;
+
+  @Inject
+  public VisibilitySystem() {
+    super();
+  }
+
+  private Bits obtainNewChangeSubscribers(int entityId) {
+    var subscribers = entityToNewChangeSubscribers.get(entityId);
+    if (subscribers == null) {
+      subscribers = new Bits();
+      entityToNewChangeSubscribers.put(entityId, subscribers);
+    }
+    return subscribers;
+  }
+
+  @Override
+  protected void begin() {
+    log.info("process VisibilitySystem ");
+    entityToNewChangeSubscribers.clear();
+  }
+
+  @Override
+  protected void process(int perceiver) {
+    var coordinates = coordinatesMapper.get(perceiver);
+    var sightlineRadius = sightlineSubscribersMapper.get(perceiver).getSightlineRadius();
+    var sightlineSubscribers = sightlineSubscribersMapper.get(perceiver).getClients();
+    log.info(coordinates.toString());
+    for (int i = 0; i < allThatCanBePerceived.getEntities().size(); i++) {
+      var perceivableCoords = coordinatesMapper.get(allThatCanBePerceived.getEntities().get(i));
+      var dst2 = Math.pow(coordinates.getX() - perceivableCoords.getX(), 2) + Math.pow(coordinates.getY() - perceivableCoords.getY(), 2);
+      if (dst2 <= sightlineRadius * sightlineRadius) {
+        var changeSubscribers = obtainNewChangeSubscribers(allThatCanBePerceived.getEntities().get(i));
+        changeSubscribers.or(sightlineSubscribers);
+      }
```

Comment:
Tu jest ładne miejsce na jakiś algorytm, który podzieli te koordynaty na sektory np. QuadTree. Chciałem użyć, ale nie wiem, ta wersja, którą znalazłem była jakaś zabugowana i przez to mi nie działało.

Reply:
I myślę, że to może być akapitem w inżynierce, a jak się wprowadzi algorytm to z 2 stronami.

Reply:
To trochę nietrafiony pomysł dla samych pól, bo one są ustawione równomiernie. 

Ale dla jednostek pozwoliłoby to na grupowanie przestrzeni w jakieś sektory, żeby nie musieć porównywać wszystkiego do wszystkiego tak jak tutaj robię, tylko na odrzucenie części, która "na pewno" nie jest blisko danych koordynatów.

Na pewno da się jakoś zastosować własności siatki do określania widoczności, ale nie chciało mi się aż tak. 

Obecnie z tym jak to sprawdzam jest problem, bo nie daje mi takiego fajnego kształtu, tylko jest nierównomiernie, ale też nie miałem głowy do takiego szczegółu, skupiałem się żeby to jako-tako działało. Dlatego tyle commitów z "fixes". Obecnie to mam nadzieję, że jedyne bugi są z powodu złej obsługi wielowątkowości i mam trooochę pojęcie jak to naprawić, ale nie zacząłem tego konkretnie jeszcze pisać.